### PR TITLE
Refactor testsuite to be more aligned with Gateway API

### PR DESCRIPTION
**Pull request by pehala**
* New structure contains these fixture/parts
  * `backend()` + `Httpbin` class - Backend to which the requests should get routed
  * `gateway()` fixture + `Gateway` class - Equivalent of Gateway object in Gateway API
  * `route()` fixture + `GatewayRoute` class - Equivalent of HTTPRoute object in Gateway API
  * `exposer()` fixture + `Exposer` class - Responsible for exposing hostname to be accessible outside of a cluster
     * For Service Protection only, it is be `OpenShiftExposer`, which uses OpenShift routes
     * For MGC it is `DNSPolicyExposer`, which relies on DNSPolicy to do the heavy lifting and expose the actual hostname
  * `hostname()` fixture and `Hostname` class - Exposed Hostname, for which we can create HTTP client to send requests
* Redone pure Envoy deployment
  * Make to emulate how Gateway API works - it is now slower due to a few rollouts needed
  * Update config instead of having to create new template file for each configuration (e.g. Wristband and TLS)
* Simplified some tests due to the new structure

### CODE REVIEW

File: testsuite/openshift/objects/envoy/tls.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+from typing import TYPE_CHECKING
+
+import yaml
+
+from . import Envoy, EnvoyConfig
+
+if TYPE_CHECKING:
+    from ...client import OpenShiftClient
+
+TLS_TRANSPORT = """
```
Comment (jsmolar) [inline testsuite/openshift/objects/envoy/tls.py:12]: Why you moved this from yaml file to constant?

Comment (pehala) [reply testsuite/openshift/objects/envoy/tls.py:12]: Les duplicatiom, the file contained all the resources, this only relevant sections

File: testsuite/objects/hostname.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    """
+
+    @abstractmethod
+    def client(self, **kwargs) -> Client:
+        """Return Httpx client for the requests to this backend"""
+
+    @property
+    @abstractmethod
+    def hostname(self):
+        """Returns full hostname in string form associated with this object"""
```
[the remark this answers is not in the document]
Comment (pehala) [reply testsuite/objects/hostname.py:23]: I see no problem using property with abstractmethod, this is also not a first time we used it, no problems so far

**Comment by jsmolar** [comment]
move `objects.__init__` to openshift.

File: testsuite/tests/mgc/conftest.py [hunk 3]
```diff
@@ -31,13 +32,13 @@ def spokes(testconfig):
 
 
 @pytest.fixture(scope="module")
-def upstream_gateway(request, openshift, blame, hostname, module_label):
+def upstream_gateway(request, openshift, blame, base_domain, module_label) -> MGCGateway:
```
[the remark this answers is not in the document]
Comment (jsmolar) [reply testsuite/tests/mgc/conftest.py:35]: TODO for future PR (maybe, just maybe for this one). Rename this fixture or add doc that better describes "hub gateway, which is reflected to spokes"

File: testsuite/tests/kuadrant/authorino/operator/test_wildcard.py [hunk 4]
```diff
@@ -6,13 +6,19 @@
...
+    """Set route for wildcard domain"""
+    route.add_hostname(wildcard_domain)
+    route.remove_hostname(hostname.hostname)
+    return route
+
+
 # pylint: disable = unused-argument
 @pytest.fixture(scope="module")
-def authorization(authorino, blame, openshift, module_label, wildcard_domain):
+def authorization(authorino, blame, route, openshift, module_label, wildcard_domain):
```
Comment (jsmolar) [inline testsuite/tests/kuadrant/authorino/operator/test_wildcard.py:19]: `wildcard_domain` is no longer needed

File: testsuite/openshift/objects/auth_config/__init__.py [hunk 5]
```diff
@@ -42,19 +41,19 @@ def create_instance(
...
             "apiVersion": "authorino.kuadrant.io/v1beta2",
             "kind": "AuthConfig",
             "metadata": {"name": name, "namespace": openshift.project, "labels": labels},
-            "spec": {"hosts": hostnames or [route.hostname]},  # type: ignore
+            "spec": {"hosts": []},  # type: ignore
         }
-
-        return cls(model, context=openshift.context)
+        obj = cls(model, context=openshift.context)
+        route.add_auth_config(obj)
```
[the remark this answers is not in the document]
Comment (pehala) [reply testsuite/openshift/objects/auth_config/__init__.py:55]: That is usually called in the AuthConfig builder, which I cannot do in here, due to it being created from a template

File: testsuite/tests/kuadrant/authorino/dinosaur/conftest.py [hunk 6]
```diff
@@ -95,32 +96,38 @@ def commit(authorization):
...
             "TERMS_AND_CONDITIONS": terms_and_conditions("false"),
             "CLUSTER_INFO": cluster_info(rhsso.client_name),
             "RESOURCE_INFO": resource_info("123", rhsso.client_name),
         },
     )
+    with openshift.context:
+        auth = selector.object(cls=AuthConfig)
 
-    request.addfinalizer(auth.delete)
+    route.add_auth_config(auth)
```
[the remark this answers is not in the document]
Comment (pehala) [reply testsuite/tests/kuadrant/authorino/dinosaur/conftest.py:130]: Added it

**Review by averevki — DISMISSED** [review]
LGTM

**Review by jsmolar — APPROVED** [review]

**Pull request merged**